File: variants/javascript/rca/engine.js
///////////////////////////////////////////////////////////////////////////////
// Regulated Cell Architecture (RCA) - JavaScript Variant
// Engine
//
// Micro-kernel space (Loop Engine privilege only):
// Apply returned outputs to ctx.
///////////////////////////////////////////////////////////////////////////////

const { createDataPlane, createActivityInfo, createSystemData } = require("./data");
const { State, Mode, createControlPlane } = require("./control");
const { NUM_CELLS, createCell, Task } = require("./cell");
const { buildTasks, step } = require("./thread");

///////////////////////////////////////////////////////////////////////////////
// Status printing helpers
///////////////////////////////////////////////////////////////////////////////

function printControl(ctl) {
  console.log("  Control:");
  console.log(`    state: ${ctl.state}`);
  console.log(`    mode:  ${ctl.mode}`);
}

function printData(dp) {
  console.log("  Data:");
  console.log(`    cells.count: ${dp.cells.count}`);
  console.log(`    activity:    "${dp.activity.description}"`);
}

function printCellData(cd) {
  console.log("  Effect:");
  switch (cd.tag) {
    case "none":
      console.log("    None");
      break;
    case "byte":
      console.log(`    Byte(${cd.value})`);
      break;
  }
}

///////////////////////////////////////////////////////////////////////////////
// (1) Engine
///////////////////////////////////////////////////////////////////////////////

// Status: FREEZE
function giveDefault() {
  return {
    ctx: createDataPlane(NUM_CELLS),
    ctl: createControlPlane(),
    sys: createSystemData(),
  };
}

function printInitStatus(engine) {
  console.log("\n>>>");
  printControl(engine.ctl);
  console.log("");
  printData(engine.ctx);
  console.log("");
}

function printRunningStatus(engine, efx) {
  console.log("\n>>>");
  printControl(engine.ctl);
  console.log("");
  printCellData(efx);
  console.log("");
  printData(engine.ctx);
  console.log("<<<");
}

function printShutdownStatus(engine) {
  console.log("\n>>>");
  printControl(engine.ctl);
  console.log("");
  printData(engine.ctx);
  console.log("");
}

///////////////////////////////////////////////////////////////////////////////
// Engine run
///////////////////////////////////////////////////////////////////////////////

function tryRunEngine(engine) {
  printInitStatus(engine);

  const currentThread = buildTasks([
    createCell(0, Task.DEFAULT),
    createCell(1, Task.DOUBLE_VALUE),
  ]);

  engine.ctl.state = State.HALT;
  printInitStatus(engine);

  engine.ctl.state = State.IDLE;
  printInitStatus(engine);

  engine.ctl.state = State.RUNNING;
  printInitStatus(engine);

  for (;;) {
    switch (engine.ctl.state) {
      case State.RUNNING: {
        const effect = step(currentThread, engine.ctx);
        engine.ctx.activity = effect.activity;

        if (engine.ctl.mode === Mode.DEBUG) {
          printRunningStatus(engine, effect.handoff);
        }

        if (effect.finished) {
          engine.ctl.state = State.SHUTDOWN;
          engine.ctx.activity = createActivityInfo();
          printShutdownStatus(engine);
          return 0;
        }
        break;
      }

      default: {
        engine.ctl.state = State.SHUTDOWN;
        printShutdownStatus(engine);
        return 0;
      }
    }
  }
}

///////////////////////////////////////////////////////////////////////////////
// (2) Add custom engine functions here
///////////////////////////////////////////////////////////////////////////////

module.exports = {
  giveDefault,
  tryRunEngine,
};

File: variants/java/rca/Engine.java
///////////////////////////////////////////////////////////////////////////////
// Regulated Cell Architecture (RCA) - Java Variant
// Engine
//
// Micro-kernel space (Loop Engine privilege only):
// Apply returned outputs to ctx.
///////////////////////////////////////////////////////////////////////////////

package rca;

///////////////////////////////////////////////////////////////////////////////
// Status printing helpers
///////////////////////////////////////////////////////////////////////////////

class Printer {
    static void printControl(Control ctl) {
        System.out.println("  Control:");
        System.out.println("    state: " + ctl.state.displayName());
        System.out.println("    mode:  " + ctl.mode.displayName());
    }

    static void printData(Data dp) {
        System.out.println("  Data:");
        System.out.println("    cells.count: " + dp.cells.count());
        System.out.println("    activity:    \"" + dp.activity.description() + "\"");
    }

    static void printCellData(CellData cd) {
        System.out.println("  Effect:");
        switch (cd) {
            case CellData.None n   -> System.out.println("    None");
            case CellData.Byte b   -> System.out.println("    Byte(" + b.value() + ")");
        }
    }
}

///////////////////////////////////////////////////////////////////////////////
// (1) Engine
///////////////////////////////////////////////////////////////////////////////

// Status: FREEZE
public class Engine {

    Data ctx;
    Control ctl;
    SystemData sys;

    Engine(Data ctx, Control ctl, SystemData sys) {
        this.ctx = ctx;
        this.ctl = ctl;
        this.sys = sys;
    }

    public static Engine giveDefault() {
        return new Engine(
            new Data(Cell.NUM_CELLS),
            new Control(),
            SystemData.defaults()
        );
    }

    void printInitStatus() {
        System.out.println("\n>>>");
        Printer.printControl(ctl);
        System.out.println();
        Printer.printData(ctx);
        System.out.println();
    }

    void printRunningStatus(CellData efx) {
        System.out.println("\n>>>");
        Printer.printControl(ctl);
        System.out.println();
        Printer.printCellData(efx);
        System.out.println();
        Printer.printData(ctx);
        System.out.println("<<<");
    }

    void printShutdownStatus() {
        System.out.println("\n>>>");
        Printer.printControl(ctl);
        System.out.println();
        Printer.printData(ctx);
        System.out.println();
    }

    ///////////////////////////////////////////////////////////////////////////
    // Engine run
    ///////////////////////////////////////////////////////////////////////////

    public int tryRunEngine() {
        printInitStatus();

        var currentThread = RcaThread.buildTasks(
            new Cell[] {
                new Cell(0, Task.DEFAULT),
                new Cell(1, Task.DOUBLE_VALUE),
            },
            null
        );

        ctl.state = State.HALT;
        printInitStatus();

        ctl.state = State.IDLE;
        printInitStatus();

        ctl.state = State.RUNNING;
        printInitStatus();

        while (true) {
            switch (ctl.state) {
                case RUNNING -> {
                    var effect = currentThread.step(ctx);
                    ctx.activity = effect.activity();

                    if (ctl.mode == Mode.DEBUG) {
                        printRunningStatus(effect.handoff());
                    }

                    if (effect.finished()) {
                        ctl.state = State.SHUTDOWN;
                        ctx.activity = ActivityInfo.defaults();
                        printShutdownStatus();
                        return 0;
                    }
                }

                default -> {
                    ctl.state = State.SHUTDOWN;
                    printShutdownStatus();
                    return 0;
                }
            }
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    // (2) Add custom engine methods here
    ///////////////////////////////////////////////////////////////////////////
}
